### PR TITLE
Prepend SageMaker Notebook Instance version ID to boto3 User Agent

### PR DESCRIPTION
**Pull request by mckev-amazon**
SageMaker Notebook Instances provide the version ID as a file on a
Notebook Instance.  In order to enhance version adoption from within
SageMaker, provide an extra segment to the user_agent supplied to the
boto3 ClientConfig.

*Issue #, if available:*

*Description of changes:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

- [X] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have updated the [changelog](https://github.com/aws/sagemaker-python-sdk/blob/master/CHANGELOG.rst) with a description of my changes (if appropriate)
- [ ] I have updated any necessary [documentation](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/user_agent.py [hunk 1]
```diff
@@ -25,9 +25,23 @@
 PYTHON_VERSION = '{}.{}.{}'.format(sys.version_info.major, sys.version_info.minor, sys.version_info.micro)
 
 
-def prepend_user_agent(client):
+def determine_prefix():
     prefix = 'AWS-SageMaker-Python-SDK/{} Python/{} {}/{} Boto3/{} Botocore/{}'\
         .format(SDK_VERSION, PYTHON_VERSION, OS_NAME, OS_VERSION, boto3.__version__, botocore.__version__)
+
+    try:
+        with open('/etc/opt/ml/sagemaker-notebook-instance-version.txt') as sagemaker_nbi_file:
```
Comment (nadiaya) [inline src/sagemaker/user_agent.py]: Add tests.

File: src/sagemaker/user_agent.py [hunk 2]
```diff
@@ -25,9 +25,23 @@
 PYTHON_VERSION = '{}.{}.{}'.format(sys.version_info.major, sys.version_info.minor, sys.version_info.micro)
 
 
-def prepend_user_agent(client):
+def determine_prefix():
     prefix = 'AWS-SageMaker-Python-SDK/{} Python/{} {}/{} Boto3/{} Botocore/{}'\
         .format(SDK_VERSION, PYTHON_VERSION, OS_NAME, OS_VERSION, boto3.__version__, botocore.__version__)
+
+    try:
```
Comment (laurenyu) [inline src/sagemaker/user_agent.py]: would it be better to use `os.path.exists` rather than a try block?

File: src/sagemaker/user_agent.py [hunk 3]
```diff
@@ -25,9 +25,23 @@
...
+        with open('/etc/opt/ml/sagemaker-notebook-instance-version.txt') as sagemaker_nbi_file:
+            prefix = 'AWS-SageMaker-Notebook-Instance/{} {}'.format(sagemaker_nbi_file.read().strip(), prefix)
+    except IOError:
```
Comment (nadiaya) [inline src/sagemaker/user_agent.py]: Should we make the error even broader to minimize any possibility of it affecting customers?

Comment (mckev-amazon) [reply src/sagemaker/user_agent.py]: I don't think so, if something besides an IOError is thrown it should fail pretty hard, IMO.

File: src/sagemaker/user_agent.py (hunk 2 above)
Comment (mckev-amazon) [reply src/sagemaker/user_agent.py]: No, to do so wouldn't be pythonic and also exposes a race condition (the file could exist in line 1 but not exist or isn't accessible by the time line 2 is run).  Doing so via a try...except block is idomatic in Python.  This is called "EAFP" - see a good example here:

https://docs.python.org/3.4/library/os.html?highlight=eafp#os.access

File: src/sagemaker/user_agent.py (hunk 1 above)
Comment (mckev-amazon) [reply src/sagemaker/user_agent.py]: Yep, I can do that.  Do note that prior to this there weren't any tests on the user_agent.py class, I'll do what I can.

File: tests/unit/test_session.py [hunk 4]
```diff
@@ -139,6 +145,57 @@ def test_delete_endpoint(boto_session):
...
+    assert 'AWS-SageMaker-Python-SDK' in sess.sagemaker_runtime_client._client_config.user_agent
+    assert 'AWS-SageMaker-Notebook-Instance' not in sess.sagemaker_client._client_config.user_agent
+    assert 'AWS-SageMaker-Notebook-Instance' not in sess.sagemaker_runtime_client._client_config.user_agent
+
+
+def patch_open(*args, **kwargs):
+    """Depending on the current Python version, patch the appropriate built-in open() function."""
+
+    try:
+        # Python 3.x
```
[the remark this answers is not in the document]
Comment (mckev-amazon) [reply tests/unit/test_session.py]: TIL, thanks for pointing this out.  Will make the change.

**Pull request merged**